components/apimgt/org.wso2.carbon.apimgt.rest.api.util/src/main/java/org/wso2/carbon/apimgt/rest/api/util/impl/OAuthJwtAuthenticatorImpl.java

```diff
@@ -0,0 +1,349 @@
+import org.wso2.carbon.context.PrivilegedCarbonContext;
+import org.wso2.carbon.apimgt.impl.APIConstants.JwtTokenConstants;
+import org.wso2.carbon.apimgt.rest.api.common.APIMConfigUtil;
+import org.wso2.carbon.apimgt.rest.api.util.MethodStats;
+import org.wso2.carbon.apimgt.rest.api.util.authenticators.AbstractOAuthAuthenticator;
+
+import static org.wso2.carbon.apimgt.rest.api.common.APIMConfigUtil.getRestApiJWTAuthAudiences;
+
+public class OAuthJwtAuthenticatorImpl extends AbstractOAuthAuthenticator {
+
+    private static final Log log = LogFactory.getLog(OAuthJwtAuthenticatorImpl.class);
+    private static final String SUPER_TENANT_SUFFIX =
+            APIConstants.EMAIL_DOMAIN_SEPARATOR + MultitenantConstants.SUPER_TENANT_DOMAIN_NAME;
+    private boolean isRESTApiTokenCacheEnabled;
+    private JWTValidator jwtValidator;
+    private Map<String, TokenIssuerDto> tokenIssuers;
+    private java.util.Map<String, String> audiencesList;
+
+    public OAuthJwtAuthenticatorImpl() {
+        jwtValidator = new JWTValidatorImpl();
+        tokenIssuers = getTokenIssuers();
+        audiencesList = getRestApiJWTAuthAudiences();
+    }
+
+    /**
+     * @param message cxf message to be authenticated
+     * @return true if authentication was successful else false
+     */
+    @Override
+    public boolean authenticate(Message message) throws APIManagementException {
+
+        RESTAPICacheConfiguration cacheConfiguration = APIUtil.getRESTAPICacheConfig();
+        isRESTApiTokenCacheEnabled = cacheConfiguration.isTokenCacheEnabled();
+        String accessToken = RestApiUtil.extractOAuthAccessTokenFromMessage(message,
+                RestApiConstants.REGEX_BEARER_PATTERN, RestApiConstants.AUTH_HEADER_NAME);
+
+        if (StringUtils.countMatches(accessToken, APIConstants.DOT) != 2) {
+            log.error("Invalid JWT token. The expected token format is <header.payload.signature>");
+            return false;
+        }
+        try {
+            SignedJWTInfo signedJWTInfo = getSignedJwt(accessToken);
+            String jwtTokenIdentifier = getJWTTokenIdentifier(signedJWTInfo);
+            String jwtHeader = signedJWTInfo.getSignedJWT().getHeader().toString();
+            URL basePath = new URL(message.get(APIConstants.BASE_PATH).toString());
+
+            //Validate token
+            log.info("Starting JWT token validation");
+            JWTValidationInfo jwtValidationInfo =
+                    validateJWTToken(signedJWTInfo, jwtTokenIdentifier, accessToken, jwtHeader, basePath);
+            if (jwtValidationInfo != null) {
+                if (jwtValidationInfo.isValid()) {
+                    if (isRESTApiTokenCacheEnabled) {
+                        getRESTAPITokenCache().put(jwtTokenIdentifier, jwtValidationInfo);
+                    }
+                    //Validating scopes
+                    log.info("JWT token validation success. starting scope validation");
```

Comment:
Same here

Reply:
CPU cost for getting the masked token can be slightly high because the JWT tokens are bit lengthy, so lets use log.isDebugEnabled in all places.

Reply:
getMaskedToken(jwtHeader) can be also called once and stored in a variable in the method level and reuse whenever needed.